### PR TITLE
[MLIR][Math] add canonicalize-f32-promotion pass

### PR DESCRIPTION
**Pull request by crazydemo**
The current `legalize-to-f32` pass does f32 promotion for every op belonging to the illegal op list. Once there are some consecutive illegal ops, `legalize-to-f32` will insert redundant `arith.truncf` and `arith.extf` pairs between the illegal ops.

This pass is to eliminate the redundant truncf/extf pairs to improve performance. However, this pass may introduce numerical difference as the `f32->bf16` rounding is eliminated.

### CODE REVIEW

**Comment by crazydemo** [comment]
@ZhennanQin @ciyongch @Menooker Could you please help review this PR?

File: mlir/include/mlir/Dialect/Math/Transforms/Passes.td [hunk 1]
```diff
@@ -36,4 +36,47 @@ def MathLegalizeToF32 : Pass<"math-legalize-to-f32"> {
...
+    This pass is to eliminate the redundant truncf/extf pairs.
+
+    Example:
+
+    ```mlir
+    // the initial func
+    func.func @bf16_sin_vector(%arg0: vector<32xbf16>) -> vector<32xbf16> {
+        %0 = math.absf %arg0 : vector<32xbf16>
+        %1 = math.sin %0 : vector<32xbf16>
+        return %2 : vector<32xbf16>
```
Comment (ciyongch) [inline mlir/include/mlir/Dialect/Math/Transforms/Passes.td]: `return 1%`?

Comment (crazydemo) [reply mlir/include/mlir/Dialect/Math/Transforms/Passes.td]: thanks for pointing out, fixed the typo.

File: mlir/test/Dialect/Math/canonicalize-f32-promotion.mlir [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+// CHECK: [[EXTF:%.+]] = arith.extf [[ARG0]]
+// CHECK: [[ABSF:%.+]] = math.absf [[EXTF]]
+// CHECK: [[SIN:%.+]] = math.sin [[ABSF]]
+// CHECK: [[TRUNCF:%.+]] = arith.truncf [[SIN]]
+// CHECK: return [[TRUNCF]] : vector<32x32x32xf16>
+func.func @f16_sin_vector(%arg0: vector<32x32x32xf16>) -> vector<32x32x32xf16> {
+  %0 = math.absf %arg0 : vector<32x32x32xf16>
+  %1 = math.sin %0 : vector<32x32x32xf16>
+  return %1 : vector<32x32x32xf16>
+}
```
Comment (ciyongch) [inline mlir/test/Dialect/Math/canonicalize-f32-promotion.mlir]: Can you also add some cases includes **branching**?

File: mlir/lib/Dialect/Math/Transforms/CanonicalizeF32Promotion.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+using namespace mlir;
+
+namespace {
+
+struct CanonicalizeF32PromotionRewritePattern final
+    : OpRewritePattern<arith::ExtFOp> {
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(arith::ExtFOp op,
+                                PatternRewriter &rewriter) const final {
+    if (auto innertruncop = op.getOperand().getDefiningOp<arith::TruncFOp>()) {
```
Comment (ciyongch) [inline mlir/lib/Dialect/Math/Transforms/CanonicalizeF32Promotion.cpp]: How about the case as below:
```mlir
        %0 = math.absf %arg0 : vector<32xbf16>
        %1 = math.sin %0 : vector<32xbf16>
        %2 = math.cos %0: vector<32xbf16>
        %3 = arith.addf %1, %2
```

Comment (crazydemo) [reply mlir/lib/Dialect/Math/Transforms/CanonicalizeF32Promotion.cpp]: Included this case as branch case in [canonicalize-f32-promotion.mlir](https://github.com/crazydemo/llvm-project/pull/1/files#diff-23020567471b87d9fa12d9e9be9851417f9abb82d394dfdb4d30328e6655633e). Based on the current upstream status, we will get code like below. bf16 legalization is not applied to `arith.addf`, as the current upstream does not add this op into the `illegal op list`. 
```
func.func @bf16_branch_vector(%arg0: vector<32x32x32xbf16>) -> vector<32x32x32xbf16> {
    %0 = arith.extf %arg0 : vector<32x32x32xbf16> to vector<32x32x32xf32>
    %1 = math.absf %0 : vector<32x32x32xf32>
    %2 = math.sin %1 : vector<32x32x32xf32>
    %3 = arith.truncf %2 : vector<32x32x32xf32> to vector<32x32x32xbf16>
    %4 = math.cos %1 : vector<32x32x32xf32>
    %5 = arith.truncf %4 : vector<32x32x32xf32> to vector<32x32x32xbf16>
    %6 = arith.addf %3, %5 : vector<32x32x32xbf16>
    return %6 : vector<32x32x32xbf16>
}
```
I will have another PR submitted to graph compiler repo, which enables graph compiler specific bf16 legalization pass (has our own illegal op list, and contain the machine info). Graph compiler specific bf16 legalization full process will generate the below ir.
```
func.func @bf16_branch_vector(%arg0: vector<32x32x32xbf16>) -> vector<32x32x32xbf16> {
    %0 = arith.extf %arg0 : vector<32x32x32xbf16> to vector<32x32x32xf32>
    %1 = math.absf %0 : vector<32x32x32xf32>
    %2 = math.sin %1 : vector<32x32x32xf32>
    %3 = math.cos %1 : vector<32x32x32xf32>
    %4 = arith.addf %2, %3 : vector<32x32x32xf32>
    %5 = arith.truncf %4 : vector<32x32x32xf32> to vector<32x32x32xbf16>
    return %5 : vector<32x32x32xbf16>
}
```

File: mlir/lib/Dialect/Math/Transforms/CanonicalizeF32Promotion.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
+//===- CanonicalizeF32Promotion.cpp - Remove redundant extf/truncf pairs
+//----------===//
```
Comment (Menooker) [inline mlir/lib/Dialect/Math/Transforms/CanonicalizeF32Promotion.cpp]: remove "-------" to move this line to above line, for formatting?

Comment (crazydemo) [reply mlir/lib/Dialect/Math/Transforms/CanonicalizeF32Promotion.cpp]: thanks for the comment. removed "-------"

File: mlir/include/mlir/Dialect/Math/Transforms/Passes.td [hunk 5]
```diff
@@ -36,4 +36,47 @@ def MathLegalizeToF32 : Pass<"math-legalize-to-f32"> {
...
 
+def MathCanonicalizeF32Promotion : Pass<"math-canonicalize-f32-promotion"> {
+  let summary = "Eliminate redundant truncf/extf pairs";
+  let description = [{
+    `legalize-to-f32` pass does f32 promotion for every op belonging to the
+    illegal op list. Once there are some consecutive illegal ops, `legalize-to-f32`
+    will insert redundant `arith.truncf` and `arith.extf` pairs between the illegal
+    ops.
+    
+    This pass is to eliminate the redundant truncf/extf pairs.
```
Comment (ZhennanQin) [inline mlir/include/mlir/Dialect/Math/Transforms/Passes.td]: Please notice the user that this pass will introduce numerical difference as the fp32_to_bf16 rounding is eliminated.

Comment (crazydemo) [reply mlir/include/mlir/Dialect/Math/Transforms/Passes.td]: Thanks for the reminder. Added the notice.

**Review by ciyongch — APPROVED** [review]

**Comment by crazydemo** [comment]
@ZhennanQin @ciyongch @Menooker This [PR](https://github.com/llvm/llvm-project/pull/92482) has been submitted to MLIR community , and has got some comments.

**Pull request closed without merging**